### PR TITLE
[2.3] Static Code Analysis for Components

### PR DESCRIPTION
**Pull request by kalessil**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Static Code Analysis with Php Inspections (EA Extended):
    - remaining mkdir race conditions
    - continue miss-usage in switch

### CODE REVIEW

**Comment by EVODelavega** [comment]
Please use `sprintf` for exception messages, instead of concatenating variables into the message:

> Yes, using sprintf is what the coding standard should document. Can you 
submit a PR on the docs? And perhaps a PR on symfony/symfony to fix 
inlined exceptions? 

>Thanks, 
Fabien

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php [hunk 1]
```diff
@@ -116,8 +116,8 @@ public static function load($classes, $cacheDir, $name, $autoReload, $adaptive =
         }
 
         // cache the core classes
-        if (!is_dir(dirname($cache))) {
-            mkdir(dirname($cache), 0777, true);
+        if (!file_exists($cacheDir) && !@mkdir($cacheDir, 0777, true) && !is_dir($cacheDir)) {
+            throw new \RuntimeException('Class Collection Loader was not able to create a folder: '.$cacheDir);
```
Comment (iltar) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: I would call it a 'directory' instead and make sure exceptions end with a full stop, same goes for the other exceptions.

**Comment by Tobion** [comment]
According to https://github.com/twigphp/Twig/issues/1944 this is not fully safe against race conditions either. And now we might throw exceptions in placed where it worked before.

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php (hunk 1 above)
Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: Corrected

**Comment by kalessil** [comment]
@Tobion : the same as in #16292 - pattern and solution; I'm not sure if this breaks BC.

**Comment by kalessil** [comment]
Exceptions message generation changed in favor of sprintf

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php [hunk 2]
```diff
@@ -116,8 +116,8 @@ public static function load($classes, $cacheDir, $name, $autoReload, $adaptive =
         }
 
         // cache the core classes
-        if (!is_dir(dirname($cache))) {
-            mkdir(dirname($cache), 0777, true);
+        if (!is_dir($cacheDir) && !@mkdir($cacheDir, 0777, true) && !is_dir($cacheDir)) {
+            throw new \RuntimeException(sprintf('Class Collection Loader was not able to create a directory: %s', $cacheDir));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: Does this add anything? If the directory doesn't exists, writeCacheFile already throws an exception...

File: src/Symfony/Component/ClassLoader/ClassMapGenerator.php [hunk 3]
```diff
@@ -134,7 +134,7 @@ private static function findClasses($path)
                     }
 
                     if ($isClassConstant) {
-                        continue;
+                        continue 2;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: I suggest using just `break` here and also removing the `default` case which looks superfluous, isn't it?

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php [hunk 4]
```diff
@@ -48,8 +48,8 @@ public function __construct($savePath = null)
             $baseDir = ltrim(strrchr($savePath, ';'), ';');
         }
 
-        if ($baseDir && !is_dir($baseDir)) {
-            mkdir($baseDir, 0777, true);
+        if ($baseDir && !is_dir($baseDir) && !@mkdir($baseDir, 0777, true) && !is_dir($baseDir)) {
+            throw new \RuntimeException(sprintf('The Session Storage was not able to create a directory: %s', $baseDir));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php]: did anyone report the need for these exceptions? I'm sure something already fails when the dir can't be created. Did you experience any "unclear" error?

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php (hunk 2 above)
Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: Do you mean "Failed to write cache file"?

File: src/Symfony/Component/ClassLoader/ClassMapGenerator.php (hunk 3 above)
Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: Break would be nicer, agree. Not having default is quite bad practice actually, because not clear what shall happen with non-handled values.

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php (hunk 4 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php]: Nope, it was no reports about this. Yep, had to support couple cases when settings were misconfigured and these exceptions pointing problematic configurations really fast.

In 99.9% of cases (I guess) exceptions will not make any harm due to default settings.

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php (hunk 2 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: Yes

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php]: Ok then , thanks for the argument :)
I suggest rewording to remove the `:` in the message: 
`The Session Storage was not able to create directory "%s".`

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php (hunk 2 above)
Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: I'd rather expect  the component to separate exceptions: because case "directory can not be created" and "file not dumped/file not renamed" needs different investigations.

File: src/Symfony/Component/Templating/Loader/CacheLoader.php [hunk 5]
```diff
@@ -69,8 +69,8 @@ public function load(TemplateReferenceInterface $template)
 
         $content = $storage->getContent();
 
-        if (!is_dir($dir)) {
-            mkdir($dir, 0777, true);
+        if (!is_dir($dir) && !@mkdir($dir, 0777, true) && !is_dir($dir)) {
+            throw new \RuntimeException(sprintf('Cache Loader was not able to create a directory: %s', $dir));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Templating/Loader/CacheLoader.php]: sometimes you have `The` in front, sometimes no. My soft preference would be without `The`, and my strong preference would be for consistency :)

File: src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php (hunk 4 above)
Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/Session/Storage/Handler/NativeFileSessionHandler.php]: Glad to share =)

Will apply wording adjustments for all exceptions.

File: src/Symfony/Component/Templating/Loader/CacheLoader.php (hunk 5 above)
Comment (kalessil) [inline src/Symfony/Component/Templating/Loader/CacheLoader.php]: Agree, will adjust.

Comment (kalessil) [inline src/Symfony/Component/Templating/Loader/CacheLoader.php]: Adjusted

File: src/Symfony/Component/HttpFoundation/HeaderBag.php [hunk 6]
```diff
@@ -52,7 +52,7 @@ public function __toString()
         foreach ($this->headers as $name => $values) {
             $name = implode('-', array_map('ucfirst', explode('-', $name)));
             foreach ($values as $value) {
-                $content .= sprintf("%-{$max}s %s\r\n", $name.':', $value);
+                $content .= sprintf("%-${max}s %s\r\n", $name.':', $value);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/HeaderBag.php]: personally, I prefer `{$max}`; it generates much cleaner tokens when parsing (think e.g. static analyzers)

Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/HeaderBag.php]: It were 4 more cases with sprintf and inline variables, as ${...} structure. Should I make adjustments to {$...} then?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/HeaderBag.php]: go for me

Comment (kalessil) [inline src/Symfony/Component/HttpFoundation/HeaderBag.php]: done

File: src/Symfony/Component/ClassLoader/ClassCollectionLoader.php [hunk 7]
```diff
@@ -116,8 +116,8 @@ public static function load($classes, $cacheDir, $name, $autoReload, $adaptive =
         }
 
         // cache the core classes
-        if (!is_dir(dirname($cache))) {
-            mkdir(dirname($cache), 0777, true);
+        if (!is_dir($cacheDir) && !@mkdir($cacheDir, 0777, true) && !is_dir($cacheDir)) {
+            throw new \RuntimeException(sprintf('Class Collection Loader was not able to create a directory "%s"', $cacheDir));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: the "a" before "directory" should be remove (below also)

Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassCollectionLoader.php]: "a"s removed in all exceptions messages from this PR

**Comment by nicolas-grekas** [comment]
:+1:
Status: reviewed

File: src/Symfony/Component/ClassLoader/ClassMapGenerator.php [hunk 8]
```diff
@@ -134,7 +134,7 @@ private static function findClasses($path)
                     }
 
                     if ($isClassConstant) {
-                        continue;
+                        break;
```
Comment (xabbuh) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: Are you sure about this change? `continue` relates to the `for` loop while `break` only leaves the `switch`.

Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: Yes, continue inside switch behaves differently in PHP than in other languages.

http://php.net/manual/en/control-structures.continue.php
```
Note: In PHP the switch statement is considered a looping structure for the purposes of continue. continue behaves like break (when no arguments are passed). If a switch is inside a loop, continue 2 will continue with the next iteration of the outer loop.
```

Comment (xabbuh) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: oh god PHP :(

**Comment by xabbuh** [comment]
:+1:

File: src/Symfony/Component/ClassLoader/ClassMapGenerator.php (hunk 8 above)
Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: I fully understand =) 

Or take in account finally-blocks introduced in 5.5 - if you get an exception inside (even handled) you can have variety of miss-behaviors of code in different PHP versions.

Comment (kalessil) [inline src/Symfony/Component/ClassLoader/ClassMapGenerator.php]: But at least I can warn people using my plugin that something might go wrong and they can concentrate on right things instead of long debugging hours.

**Comment by fabpot** [comment]
Thank you @kalessil.
